freezed_provider_value_notifier/lib/todo_list_model.dart

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 The Flutter Architecture Sample Authors. All rights reserved.
+// Use of this source code is governed by the MIT license that can be found
+// in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter/foundation.dart';
+import 'package:collection/collection.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:freezed_provider_value_notifier/models.dart';
+import 'package:todos_repository_core/todos_repository_core.dart';
+
+part 'todo_list_model.freezed.dart';
+
+enum VisibilityFilter { all, active, completed }
+
+@freezed
+abstract class TodoList with _$TodoList {
+  factory TodoList(
+    List<Todo> todos, {
+    @required VisibilityFilter filter,
+    @required bool loading,
+  }) = TodoListState;
+
+  @late
+  int get numCompleted =>
+      todos.where((Todo todo) => todo.complete).toList().length;
+
+  @late
+  bool get hasCompleted => numCompleted > 0;
+
+  @late
+  int get numActive =>
+      todos.where((Todo todo) => !todo.complete).toList().length;
+
+  @late
+  bool get hasActiveTodos => numActive > 0;
+
+  @late
+  List<Todo> get filteredTodos => todos.where((todo) {
+        switch (filter) {
+          case VisibilityFilter.active:
+            return !todo.complete;
+          case VisibilityFilter.completed:
+            return todo.complete;
+          case VisibilityFilter.all:
+          default:
+            return true;
+        }
+      }).toList();
+}
+
+extension TodoById on TodoList {
+  Todo todoById(String id) =>
+      todos.firstWhere((it) => it.id == id, orElse: () => null);
+}
+
+class TodoListController extends ValueNotifier<TodoList> {
```

Comment:
Would it make sense to move this into its own file, to help folks mentally separate the Model from the controller?

Reply:
I wouldn't recommend doing that. They are closely related and it doesn't make sense in my opinion to split them.